adminSiteServer/multiDim.ts

```diff
@@ -0,0 +1,205 @@
+            indicators: {
+                y: resolveSingleOrArrayField(view.indicators.y),
+                x: resolveSingleField(view.indicators.x),
+                size: resolveSingleField(view.indicators.size),
+                color: resolveSingleField(view.indicators.color),
+            },
+        })),
+    }
+}
+
+async function saveNewMultiDimViewChartConfig(
+    knex: db.KnexReadWriteTransaction,
+    patchConfig: GrapherInterface,
+    fullConfig: GrapherInterface
+): Promise<string> {
+    const chartConfigId = uuidv7()
+    await db.knexRaw(
+        knex,
+        `-- sql
+            INSERT INTO chart_configs (id, patch, full)
+            VALUES (?, ?, ?)
+        `,
+        [
+            chartConfigId,
+            serializeChartConfig(patchConfig),
+            serializeChartConfig(fullConfig),
+        ]
+    )
+
+    // We need to get the full config and the md5 hash from the database instead of
+    // computing our own md5 hash because MySQL normalizes JSON and our
+    // client computed md5 would be different from the ones computed by and stored in R2
+    const fullConfigMd5 = await db.knexRawFirst<
+        Pick<DbRawChartConfig, "full" | "fullMd5">
+    >(
+        knex,
+        `-- sql
+            select full, fullMd5 from chart_configs where id = ?`,
+        [chartConfigId]
+    )
+
+    await saveGrapherConfigToR2ByUUID(
+        chartConfigId,
+        fullConfigMd5!.full,
+        fullConfigMd5!.fullMd5 as Base64String
+    )
+
+    return chartConfigId
+}
+
+async function saveMultiDimConfig(
+    knex: db.KnexReadWriteTransaction,
+    slug: string,
+    config: MultiDimDataPageConfigEnriched
+) {
+    const id = await upsertMultiDimDataPage(knex, {
+        slug,
+        config: JSON.stringify(config),
+    })
+    // We need to get the full config and the md5 hash from the database instead of
+    // computing our own md5 hash because MySQL normalizes JSON and our
+    // client computed md5 would be different from the ones computed by and stored in R2
+    const result = await knex<DbPlainMultiDimDataPage>(
+        MultiDimDataPagesTableName
+    )
+        .select("config", "configMd5")
+        .where({ id })
+        .first()
+    const { config: normalizedConfig, configMd5 } = result!
+    await saveMultiDimConfigToR2(normalizedConfig, slug, configMd5)
+    return id
+}
+
+export async function createMultiDimConfig(
+    knex: db.KnexReadWriteTransaction,
+    slug: string,
+    rawConfig: MultiDimDataPageConfigRaw
+): Promise<number> {
+    const config = await resolveMultiDimDataPageCatalogPathsToIndicatorIds(
+        knex,
+        rawConfig
+    )
+    const variableConfigs = await getMergedGrapherConfigsForVariables(
+        knex,
+        uniq(config.views.map((view) => view.indicators.y[0]))
+    )
+    const enrichedViews = await Promise.all(
+        config.views.map(async (view) => {
+            const variableId = view.indicators.y[0]
+            const fullGrapherConfig = {
```

Comment:
Good catch, changed. We now get this warning: "About to merge Grapher configs with missing schema information". That's because the mdim config doesn't have a `$schema` key. Should we add it in ETL?

Reply:
Also, the last config we merge in and define inline doesn't have a `$schema`. Should we specify it there too? Hard-coding it to the latest?